homeassistant/components/heos/media_player.py

```diff
@@ -0,0 +1,195 @@
+"""
+Denon HEOS Media Player.
+"""
+
+import logging
+
+import voluptuous as vol
+
+from aioheos import AioHeosController
+
+from homeassistant.components.media_player import (PLATFORM_SCHEMA,
+                                                   MediaPlayerDevice)
+
+from homeassistant.components.media_player.const import (
+    MEDIA_TYPE_MUSIC, SUPPORT_NEXT_TRACK, SUPPORT_PAUSE, SUPPORT_PLAY,
+    SUPPORT_PLAY_MEDIA, SUPPORT_PREVIOUS_TRACK, SUPPORT_STOP,
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
+
+from homeassistant.const import (CONF_HOST, STATE_OFF, STATE_PAUSED,
+                                 STATE_PLAYING)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, async_dispatcher_send)
+
+
+REQUIREMENTS = ['aioheos==0.3.1']
+
+SUPPORT_HEOS = SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_PAUSE | \
+        SUPPORT_PLAY_MEDIA | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
+        SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST):
+    cv.string,
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discover_info=None):
+    """Set up the HEOS platform."""
+    host = config[CONF_HOST]
+
+    controller = AioHeosController(hass.loop, host, verbose=True)
+    await controller.connect(host=host)
+
+    players = controller.get_players()
+    players.sort()
+    devices = [HeosMediaPlayer(hass, p) for p in players]
+    async_add_devices(devices)
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, hass, player):
+        """Initialize."""
+        self._hass = hass
+        self._player = player
+        self._state = None
+        self._dispatcher_remove = None
+
+        self._player.state_change_callback = self.update_state
+        self._player.request_update()
```

Comment:
We don't want side-effects in `__init__`.  Add a `async_update(self)` method to the class and move this line to there.  Then pass `True` as the second parameter to `async_add_devices` in `async_setup_platform` which will cause `async_update` to be called when the entity is added to HASS.